File: adapters/android/starbidz-core/src/main/kotlin/io/starbidz/core/Starbidz.kt
package io.starbidz.core

import android.content.Context

object Starbidz {
    private var config: StarbidConfig? = null
    private var client: StarbidzClient? = null

    @Volatile
    private var isInitialized = false

    fun initialize(context: Context, config: StarbidConfig) {
        if (isInitialized) return

        synchronized(this) {
            if (isInitialized) return

            this.config = config
            this.client = StarbidzClient(config)
            isInitialized = true
        }
    }

    fun isInitialized(): Boolean = isInitialized

    fun getConfig(): StarbidConfig {
        checkInitialized()
        return config!!
    }

    internal fun getClient(): StarbidzClient {
        checkInitialized()
        return client!!
    }

    suspend fun requestBid(
        context: Context,
        placementId: String,
        format: AdFormat,
        width: Int? = null,
        height: Int? = null
    ): AdResult {
        checkInitialized()
        return client!!.requestBid(context, placementId, format, width, height)
    }

    suspend fun trackImpression(bidId: String, placementId: String) {
        client?.trackEvent("impression", bidId, placementId)
    }

    suspend fun trackClick(bidId: String, placementId: String) {
        client?.trackEvent("click", bidId, placementId)
    }

    suspend fun trackComplete(bidId: String, placementId: String) {
        client?.trackEvent("complete", bidId, placementId)
    }

    private fun checkInitialized() {
        if (!isInitialized) {
            throw IllegalStateException("Starbidz SDK is not initialized. Call Starbidz.initialize() first.")
        }
    }
}

File: adapters/ios/StarbidCore/Sources/Starbidz.swift
import Foundation

public final class Starbidz {
    public static let shared = Starbidz()

    private var config: StarbidConfig?
    private var client: StarbidClient?
    private var _isInitialized = false

    private init() {}

    public var isInitialized: Bool {
        return _isInitialized
    }

    public func initialize(config: StarbidConfig) {
        guard !_isInitialized else { return }

        self.config = config
        self.client = StarbidClient(config: config)
        self._isInitialized = true
    }

    public func getConfig() -> StarbidConfig? {
        return config
    }

    public func requestBid(
        placementId: String,
        format: AdFormat,
        width: Int? = nil,
        height: Int? = nil,
        completion: @escaping (AdResult) -> Void
    ) {
        guard _isInitialized, let client = client else {
            completion(.error("Starbidz SDK is not initialized. Call Starbidz.shared.initialize() first.", -1))
            return
        }

        client.requestBid(
            placementId: placementId,
            format: format,
            width: width,
            height: height,
            completion: completion
        )
    }

    public func trackImpression(bidId: String, placementId: String) {
        client?.trackEvent(eventType: "impression", bidId: bidId, placementId: placementId)
    }

    public func trackClick(bidId: String, placementId: String) {
        client?.trackEvent(eventType: "click", bidId: bidId, placementId: placementId)
    }

    public func trackComplete(bidId: String, placementId: String) {
        client?.trackEvent(eventType: "complete", bidId: bidId, placementId: placementId)
    }
}
